Decouple SSF spec from OAuth2 scopes (#101)
openid/sharedsignals: openid-sharedsignals-framework-1_0.md

    date: May 2021
    target: https://datatracker.ietf.org/doc/html/draft-ietf-secevent-subject-identifiers
    title: Subject Identifiers for Security Event Tokens

>   PROTECTEDSERVERMETADATA:

Reviewer: OPRM may be a shorter acronym for this.

> OPTIONAL. A list of complex type that specifies supported
  authorization scheme properties defined in {{authorization-scheme}}.  To enable seamless discovery of
  configurations, the service provider SHOULD, with the appropriate
  security considerations, make the authorization_schemes attribute
  publicly accessible without prior authentication.

TODO: consider adding a IANA Registry for metadata, similar to Section 7.1.1 of
{{RFC8414}}. This would allow other specs to add to the metadata.

### Authorization scheme {#authorization-scheme}
SSF is an HTTP based signals sharing framework. It is agnostic to the authentication and authorization schemes used to secure stream configuration APIs. It does not provide any SSF specific authentication and authorization schemes but relies on the cooperating parties mutual security considerations. Authorization scheme section of the metadata, providers discovery information related to the transmitter's stream management APIs.

spec_urn  

> REQUIRED. A URN that describes the specification of the protocol being used.

The receiver will call the transmitter APIs by providing appropriate credentials as per the `spec_urn`.

> If the `spec_urn` in Authorization scheme is `urn:ietf:rfc6749`

Reviewer: Suggested text: "The following is a non-normative example of an Authorization Server Descriptor", but you should probably add an equivalent line for the example you are providing"

> OPTIONAL. An array of member names in a Complex Subject which, if present in
  a Subject Member in an event, MUST be interpreted by a Receiver.

authorization_schemes

> OPTIONAL. A list of complex type that specifies supported
  authorization scheme properties defined in {{authorization-scheme}}.  To enable seamless discovery of
  configurations, the service provider SHOULD, with the appropriate
  security considerations, make the authorization_schemes attribute
  publicly accessible without prior authentication.

TODO: consider adding a IANA Registry for metadata, similar to Section 7.1.1 of
{{RFC8414}}. This would allow other specs to add to the metadata.

### Authorization scheme {#authorization-scheme}

> SSF is an HTTP based signals sharing framework. It is agnostic to the authentication and authorization schems used to secure stream configuration APIs. It does not provide any SSF specific authentication and authorization schemes but relies on the cooperating parties mutual security considerations. Authorization scheme section of the metadata, providers disovery informaton related to the transmitter's stream management APIs.

Reviewer 2: A minor nit-pick, s/schems/schemes/, and s/informaton/information.

spec_urn  

> > A URN for a specification for the protocol being used  REQUIRED.

Reviewer: Suggestion to rewrite as: REQUIRED. A URN that describes the specification of the protocol being used.

> OPTIONAL. An array of member names in a Complex Subject which, if present in
  a Subject Member in an event, MUST be interpreted by a Receiver.

authorization_schemes

> > OPTIONAL. A list of complex type that specifies supported

Reviewer 3: ```suggestion
> OPTIONAL. A list of JSON objects that specify the supported
```

  authorization scheme properties defined in {{authorization-scheme}}.  To enable seamless discovery of
  configurations, the service provider SHOULD, with the appropriate
  security considerations, make the authorization_schemes attribute
  publicly accessible without prior authentication.

TODO: consider adding a IANA Registry for metadata, similar to Section 7.1.1 of
{{RFC8414}}. This would allow other specs to add to the metadata.

### Authorization scheme {#authorization-scheme}

> SSF is an HTTP based signals sharing framework. It is agnostic to the authentication and authorization schemes used to secure stream configuration APIs. It does not provide any SSF specific authentication and authorization schemes but relies on the cooperating parties' mutual security considerations. Authorization scheme section of the metadata, providers discovery information related to the transmitter's stream management APIs.

Reviewer 3: ```suggestion
SSF is an HTTP based signals sharing framework. It is agnostic to the authentication and authorization schemes used to secure stream configuration APIs. It does not provide any SSF specific authentication and authorization schemes but relies on the cooperating parties' mutual security considerations. The authorization scheme section of the metadata provides discovery information related to the transmitter's stream management APIs.
```

> spec_urn  

Reviewer: The way this PR reads is not quite how I thought we discussed it in the meeting. Apologies if I missed saying this in my earlier review of the PR. Here's what I'd like to suggest:
* Drop the "spec_urn" field, and turn the "authorization_schemes" as an array of URNs. So for example, the array will look like:
```
"authorization_scheme" : [
    "urn:ietf:rfc:6749",
    "urn:openid:oidc",
]
```

Author: @atultulshi we decided [this](https://hackmd.io/@oidf-wg-sse/wg-meeting-20230905#Decoupling-SSF-and-OAuth) on 9/5 call and the PR represents that

Reviewer: Got it, thanks

> OPTIONAL. An array of member names in a Complex Subject which, if present in
  a Subject Member in an event, MUST be interpreted by a Receiver.

> authorization_schemes

Reviewer: I'm trying to understand if this is the right way to add protected server metadata. Based on my understanding of the PROTECTEDSERVERMETADATA draft and the example given here: [Protected Resource Metadata Request](https://www.ietf.org/archive/id/draft-jones-oauth-resource-metadata-04.html#name-protected-resource-metadata-) and [Protected Resource Metadata Response](https://www.ietf.org/archive/id/draft-jones-oauth-resource-metadata-04.html#name-protected-resource-metadata-r), I feel we do not need to specify a new field in our metadata. Instead, we should note that implementations can serve a PROTECTEDSERVERMETADATA URL (i.e. /.well-known/oauth-protected-resource) and fields like "authorization_servers" and "scopes_provided" in that metadata to support OAuth.

Author: Not sure I am following your point. 

The idea behind adding `authorization_schemes` here was to hint the receivers what to expect. 

> fields like "authorization_servers" and "scopes_provided" in that metadata to support OAuth.

Also, I thought we agreed on decoupling SSF spec from OAuth details like "authorization_servers" and "scopes_provided", as it could be discovered using [Protected Resource Metadata Request](https://www.ietf.org/archive/id/draft-jones-oauth-resource-metadata-04.html#name-protected-resource-metadata-) and [Protected Resource Metadata Response](https://www.ietf.org/archive/id/draft-jones-oauth-resource-metadata-04.html#name-protected-resource-metadata-r)

Reviewer: I think my comment was on a [previous commit](https://github.com/openid/sharedsignals/pull/101/commits/b585ac8418d08d9fb80a425b5401f93ea0aa2e4d#diff-193e4a2db2f645f1b4f19c4a9f8facbef29f86e122a8da4c529542750a291a1cR554) that stated this a bit differently. This specific comment can be considered as resolved for now.

> OPTIONAL. An array of member names in a Complex Subject which, if present in
  a Subject Member in an event, MUST be interpreted by a Receiver.

> authorizationSchemes

Reviewer 4: for consistency with the rest of the metadata, should this be `authorization_schemes`?

Author: Updated

authorization_schemes

> OPTIONAL. A multi-valued complex type that specifies supported
  authorization scheme properties defined in {{authorization-scheme}}.  To enable seamless discovery of
  configurations, the service provider SHOULD, with the appropriate
  security considerations, make the authorization_schemes attribute
  publicly accessible without prior authentication.

TODO: consider adding a IANA Registry for metadata, similar to Section 7.1.1 of
{{RFC8414}}. This would allow other specs to add to the metadata.

### Authorization scheme {#authorization-scheme}
SSF is an HTTP based signals sharing framework. It is agnostic to the authentication and authorization schems used to secure stream configuration APIs. It does not provide any SSF specific authentication and authorization schemes but relies on the cooperating parties mutual security considerations. Authorization scheme section of the metadata, providers disovery informaton related to the transmitter's stream management APIs. 

type

> The authorization scheme.  This specification defines the values "oauth",
  "oauth2", "oauthbearertoken" REQUIRED.

spec_urn  

> > A spec URN specifing the protocol being used  OPTIONAL.

Reviewer 4: suggested text: "A URN for a specification for the protocol being used"

> OPTIONAL. An array of member names in a Complex Subject which, if present in
  a Subject Member in an event, MUST be interpreted by a Receiver.

authorization_schemes

> OPTIONAL. A multi-valued complex type that specifies supported
  authorization scheme properties defined in {{authorization-scheme}}.  To enable seamless discovery of
  configurations, the service provider SHOULD, with the appropriate
  security considerations, make the authorization_schemes attribute
  publicly accessible without prior authentication.

TODO: consider adding a IANA Registry for metadata, similar to Section 7.1.1 of
{{RFC8414}}. This would allow other specs to add to the metadata.

### Authorization scheme {#authorization-scheme}
SSF is an HTTP based signals sharing framework. It is agnostic to the authentication and authorization schems used to secure stream configuration APIs. It does not provide any SSF specific authentication and authorization schemes but relies on the cooperating parties mutual security considerations. Authorization scheme section of the metadata, providers disovery informaton related to the transmitter's stream management APIs. 

type

> > The authorization scheme.  This specification defines the values "oauth",

Reviewer: Should we just put the URN / URI of the spec here (like we do for push and poll delivery)? That way we do not need additional fields.

Reviewer 5: OAUTH2 metadata & discovery should be optional in the core spec.  Everything can be agreed out-of-band for some implementations.   What we will need to layer ontop of the core spec is an OAUTH2 profile for SSF that defines mandatory OAUTH2 capabilities for interoperability.   This is where we can be optioned about what is mandatory.

Author: 1. @atultulshi I thought about it but may not fit in custom/proprietary schemes. Let me know any suggestions around this. Maybe custom schemes could be documented and populated used `documentation_uri`
2. @mcguinness please note that `authorization_schemes` itself is marked `OPTIONAL`, this would decouple any authZ schemes from SSF core spec

Reviewer 5: Sorry, thought `Should we just put the URN / URI of the spec here` was referring to the discovery spec vs the OAUTH2 spec.  Misread.

Author: Updated with the `spec_urn` @atultulshi

  "verification_endpoint":
    "https://tr.example.com/ssf/mgmt/verification",
  "critical_subject_members": [ "tenant", "user" ],
  "authorization_schemes": {
    [
      {

>         "type": "oauth2",

Reviewer 4: where are these types defined?

Author: Where I describe the `type` [here](https://github.com/openid/sharedsignals/pull/101/files#diff-193e4a2db2f645f1b4f19c4a9f8facbef29f86e122a8da4c529542750a291a1cR567)

Examples:

> 1. If a Receiver makes a request with an invalid token, then the

Reviewer: Since we are dropping references to OAuth, does it make sense to change this sentence to: "If a Receiver makes a request with improper authorization, then..." (i.e. not assume token based authorization)

   Transmitter MUST respond with a 401 error status.

> 2. If the Receiver presents a valid token, but the Transmitter policy

Reviewer: Suggest changing this to: "If the Receiver presents a valid authorization, but the..."

in Sections 4.5 and 4.6 of {{RFC8417}}.

### The "aud" Claim {#aud-claim}

> The "aud" claim can be a single value or an array. Values that uniquely identifies the Receiver to the

Reviewer 2: ```suggestion
The "aud" claim can be a single value (string) or an array of strings. 
```

in Sections 4.5 and 4.6 of {{RFC8417}}.

### The "aud" Claim {#aud-claim}
The "aud" claim can be a single value (string) or an array of strings. Values that 

> uniquely identifies the Receiver to the Transmitter MAY be used, if the two parties

Reviewer 3: ```suggestion
uniquely identify the Receiver to the Transmitter MAY be used, if the two parties
```

in Sections 4.5 and 4.6 of {{RFC8417}}.

### The "aud" Claim {#aud-claim}

> The "aud" claim can be a single value (string) or an array of strings. Values that 

Reviewer 4: "...can be a string or an array of strings."

I don't think "single value" is needed. Alternative would be "a single string or an array of strings"
